### PR TITLE
Followup of #817

### LINKED ISSUE #570: Denormalize t_transactions table
- Denormalize entity ids in t_transactions table
- Get rid of any other SELECT kind of queries (on entities) in parser

### PR DESCRIPTION
**Pull request by apeksharma**
**Detailed description**:
- Change `Long` to `EntityId` in `Transaction` and `Entities`. Added `EntityIdConverter` for the same.
- Use only `id` field in EntityId for `EqualsAndHashCode` since that is sufficient
- Add InvalidEntityException. Thrown for invalid entities when encoding/decoding
- Handle the case for invalid entityId (contract/topic/file id) in a failed transaction. Added test.

Signed-off-by: Apekshit Sharma <apekshit.sharma@hedera.com>

**Which issue(s) this PR fixes**:
Fixes #570 
Part of #820 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/exception/InvalidEntityException.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+public class InvalidEntityException extends RuntimeException {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/exception/InvalidEntityException.java:23]: Should extend ImporterException

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListener.java [hunk 2]
```diff
@@ -95,7 +96,13 @@ public void onItem(RecordItem recordItem) throws ImporterException {
         TransactionHandler transactionHandler = transactionHandlerFactory.create(body);
 
         long consensusNs = Utility.timeStampInNanos(txRecord.getConsensusTimestamp());
-        EntityId entityId = transactionHandler.getEntity(recordItem);
+        EntityId entityId;
+        try {
+            entityId = transactionHandler.getEntity(recordItem);
+        } catch (InvalidEntityException e) { // transaction can have invalid topic/contract/file id
+            log.error(e);
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListener.java:103]: Please log the consensusNs as well. It should probably just be at warn level as it's due to customer input and not alert worthy.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java [hunk 3]
```diff
@@ -56,6 +57,9 @@
     private final DataSource dataSource;
     private final RecordFileRepository recordFileRepository;
     private long batch_count = 0;
+    // Keeps track of entityIds seen in the current file being processed. This is for optimizing inserts into
+    // t_entities table so that insertion of node and treasury ids are not tried for every transaction.
+    private HashSet<EntityId> entityIds;
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java:62]: nit: Should use highest level interface that still meets your requirements (Dependency Inversion Principle). Recommend Collection. Also can be initialized here.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/exception/InvalidEntityException.java (hunk 1 above)
Comment (apeksharma) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/exception/InvalidEntityException.java]: Done.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListener.java (hunk 2 above)
Comment (apeksharma) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListener.java]: Done.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java (hunk 3 above)
Comment (apeksharma) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java]: Done.

**Pull request merged**